### PR TITLE
Source-dependent analysis codes with wobble data

### PR DESCRIPTION
**Pull request by SeiyaNozaki**
This PR is to add source-dependent parameters with wobble data.
Source-dependent parameters for ON events, i.e. an expected source position is pointing direction itself, is saved in `dl1/event/telescope/parameters_src_dependent/LST_LSTCam`.
One can choose `n_off_wobble` in config file, and source-dependent parameters with each rotation angle XXX is saved in `dl1/event/telescope/parameters_src_dependentXXX/LST_LSTCam`.
The source position is updated every 1000 events for the moment to save time. Computation time will be faster as discussed in #501, then I will change this condition to compute source positions every event.

To create dl2 files, dl1 standard parameters are copied in `dl2/event/telescope/parameters/LST_LSTCam`, and random forest outputs with each roration angle is saved in `dl2/event/telescope/parameters_src_dependentXXX/LST_LSTCam` without standard parameters to avoid saving duplicate data.

Any comments are welcome.

### CODE REVIEW

File: lstchain/reco/dl1_to_dl2.py
Comment (LukasNickel) [inline lstchain/reco/dl1_to_dl2.py:569]: I suppose this whole loop will get removed with the faster transformations?
It should not be necessary to iterate one event at at time, the transformations works on arrays/quantities.

File: lstchain/scripts/lstchain_add_source_dependent_parameters.py
Comment (LukasNickel) [inline lstchain/scripts/lstchain_add_source_dependent_parameters.py:62]: I think this would be easier to read if the params were in a dict or as separated on/off params

File: lstchain/scripts/lstchain_dl1_to_dl2.py
Comment (LukasNickel) [inline lstchain/scripts/lstchain_dl1_to_dl2.py:125]: Do you want to filter based on source dependent parameter as well?
It looks pretty inefficient to apply the same filters every time again, maybe split them into src independent and source dependent filters somehow? That would also avoid the `i==0` part below

File: lstchain/scripts/lstchain_dl1_to_dl2.py
Comment (LukasNickel) [inline lstchain/scripts/lstchain_dl1_to_dl2.py:149]: Why is this construct needed? Can't you just do
```
for i in dl1_keys:
    ....
```

File: lstchain/reco/dl1_to_dl2.py
Comment (maxnoe) [reply lstchain/reco/dl1_to_dl2.py:569]: Indepentent of faster or not faster transforms. This loop must be removed, as it will slow down the code quite dramatically.

File: lstchain/reco/dl1_to_dl2.py
Comment (SeiyaNozaki) [reply lstchain/reco/dl1_to_dl2.py]: Done!

File: lstchain/reco/dl1_to_dl2.py
Comment (SeiyaNozaki) [reply lstchain/reco/dl1_to_dl2.py]: Thank you for the suggestion, I modified the codes as you suggested

File: lstchain/reco/dl1_to_dl2.py
Comment (SeiyaNozaki) [reply lstchain/reco/dl1_to_dl2.py]: ok, I modified the codes to return a dictionary and use `MultiIndex` for `DataFrame`.

File: lstchain/reco/dl1_to_dl2.py
Comment (SeiyaNozaki) [reply lstchain/reco/dl1_to_dl2.py]: Done!

File: lstchain/reco/dl1_to_dl2.py
Comment (SeiyaNozaki) [reply lstchain/reco/dl1_to_dl2.py]: `off_angle` could be easy to understand?

File: lstchain/reco/dl1_to_dl2.py
Comment (SeiyaNozaki) [reply lstchain/reco/dl1_to_dl2.py]: I fully agree, removed the loop.

File: lstchain/scripts/lstchain_dl1_to_dl2.py
Comment (SeiyaNozaki) [reply lstchain/scripts/lstchain_dl1_to_dl2.py]: It was needed because an index could be changed while the loop removing the elements.
but anyway I removed this part, and now it is simple

File: lstchain/scripts/lstchain_add_source_dependent_parameters.py
Comment (SeiyaNozaki) [reply lstchain/scripts/lstchain_add_source_dependent_parameters.py]: same as DL1, I modified to use a dictionary and `MultiIndex`

File: lstchain/scripts/lstchain_dl1_to_dl2.py
Comment (SeiyaNozaki) [reply lstchain/scripts/lstchain_dl1_to_dl2.py:125]: In principle I assume source-independent parameters are used for the filtering, but source-dependent parameters could also be filtered because they are needed to apply random forest and removed if they are infinite.
Moreover if the parameters which we want to apply filtering are not in the data, an error is shown now. It means the same filtering cannot be used for each source-dep/indep parameters.
I'd like to consider more, any comments are welcome.

**Comment by rlopezcoto** [comment]
@SeiyaNozaki are you done with all the comments? 
@LukasNickel could you give another review and approve if everything is ok?

**Review by rlopezcoto — APPROVED** [review]
Now that conflicts have been solved, can this PR be merged?

**Comment by SeiyaNozaki** [comment]
> @SeiyaNozaki are you done with all the comments?
Sorry, I forgot to answer this question...
Yes, in principle I resolved all of the comments from @LukasNickel .
Please give me any other comments if you still have anything!

**Comment by chaimain** [comment]
The CI tests are failing as the branch does not have access to run the private test data. This will probably be the issue with all the other PR's as well, when the commits are not made from a branch of the main repo.
The solution currently, is to close the PR and re-open the PR from a branch of the main repo.

**Comment by maxnoe** [comment]
> The solution currently, is to close the PR and re-open the PR from a branch of the main repo.

Just noting: we don't expect any other solution soon.

**Comment by maxnoe** [comment]
@SeiyaNozaki I sent you an invitation to join the `lst-dev` team, which gives you write access to this repository (but not to the master branch).

If you have this repository setup as `upstream`, you can just

```
git push -u upstream srcdep_wobble
```

to push it here instead of to your fork.

For new pull requests, you can then  just start without forking.

**Comment by chaimain** [comment]
This should probably be told to all the other people who have open PRs and not members of the `lst-dev` team.
It will be another push to get some relevant PRs to be merged sooner

**Comment by rlopezcoto** [comment]
@chaimain yes, thanks, we can mention it in the lst-reco call on Monday in the AoB discussion, remind it to us if we do not remember, please.
For the future, there is some discussion in https://github.com/cta-observatory/ctapipe/issues/1615 to include this in the documentation to which we are pointing new developers

**Comment by rlopezcoto** [comment]
> @SeiyaNozaki I sent you an invitation to join the `lst-dev` team, which gives you write access to this repository (but not to the master branch).
> 
> If you have this repository setup as `upstream`, you can just
> 
> ```
> git push -u upstream srcdep_wobble
> ```
> 
> to push it here instead of to your fork.
> 
> For new pull requests, you can then just start without forking.

great, thanks for this. So whenever @SeiyaNozaki re-creates the PR, since it is already approved, we can merge it and make it part of next release

**Comment by rlopezcoto** [comment]
closed via #596

**Pull request closed without merging**
